### PR TITLE
Swagger update for AFS vaulted backup policy

### PR DESCRIPTION
**Pull request by masansari**
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [x] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by sjanamma** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/tieringCost/fetchTieringCost": {

Doesn't follow the RPC approved URI format.

it shoulde be /tieringCosts/tierCost/fetchTieringCost



---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:4589 in bdf58ba. [](commit_id = bdf58ba0dff0d74a1516ca6b58c2f589085a844a, deletion_comment = False)

**Comment by sjanamma** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/tieringCost/operationResults/{operationId}": {

segments need to be balanced. URI doest follow waing conventions


---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:4656 in bdf58ba. [](commit_id = bdf58ba0dff0d74a1516ca6b58c2f589085a844a, deletion_comment = False)

**Comment by sjanamma** [comment]
URI segments need to be fixed.

**Comment by jindani** [comment]
@rkmanda 
Comments for lint diff issues, can you please review and provide your feedback?

❌ **PathForResourceAction** | Location: Microsoft.RecoveryServices/stable/2023-06-01/bms.json#L4586 | RPC-Uri-V1-07

- Here the backupTieringCost is not the sub resource but kind of group of actions that are related to tiering costs. 
- Our exisiting apis for similar operations are designed this way for example 
    - /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig/vaultstorageconfig/triggerDataMove

**ResourceNameRestriction** x 3 errors for 3 apis
We have started the internal discussion on putting pattern restriction. Though we have our internal validations and it is fast operation. 

**LroErrorContent**

- We have our own custom standar cloud error, and we are already using it for all the existing apis

**PathForNestedResource** x 2 errors for 2 apis

- Here the "backupTieringCost" is not a nested resource and our existing apis follow similar patterns for getting the operation status and operation results

**Comment by jindani** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/tieringCost/operationResults/{operationId}": {
> > ```
> 
> segments need to be balanced. URI doest follow waing conventions
> 
> Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:4656 in [bdf58ba](https://github.com/Azure/azure-rest-api-specs/commit/bdf58ba0dff0d74a1516ca6b58c2f589085a844a). [](commit_id = [bdf58ba](https://github.com/Azure/azure-rest-api-specs/commit/bdf58ba0dff0d74a1516ca6b58c2f589085a844a), deletion_comment = False)

Can you give more details on what needs to be done or references?
We are already using the similar api formats for multiple apis 
For example - 
/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig/vaultstorageconfig/triggerDataMove

**Comment by jindani** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/tieringCost/fetchTieringCost": {
> > ```
> 
> Doesn't follow the RPC approved URI format.
> 
> it shoulde be /tieringCosts/tierCost/fetchTieringCost
> 
> Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:4589 in [bdf58ba](https://github.com/Azure/azure-rest-api-specs/commit/bdf58ba0dff0d74a1516ca6b58c2f589085a844a). [](commit_id = [bdf58ba](https://github.com/Azure/azure-rest-api-specs/commit/bdf58ba0dff0d74a1516ca6b58c2f589085a844a), deletion_comment = False)

We have some existing apis which follow the similar patterns,  Can we get exception for this api? and continue with current format? As tiering cost is not a sub resource. 
For example
/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig/vaultstorageconfig

File: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json [hunk 1]
```diff
@@ -8609,6 +8609,11 @@
           "format": "int32",
           "description": "Soft delete retention period in days",
           "type": "integer"
+        },
+        "vaultId": {
```
Comment (pratikjoshi14) [inline specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json:8613]: Why change in old API version?

Comment (hiaga) [reply specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-04-01/bms.json:8613]: @pratikjoshi14 this is to fix swagger correctness issue - https://openapihub.azure-devex-tools.com/amekpis/correctness/detail?errorId=655CA4E4-72C9-4E51-90A5-569420E0B8EF

**Comment by pratikjoshi14** [comment]
>     "RestrictedVaultName": {

Why is this separate path parameter needed? How would it work on existing resources?

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:13477 in e829cd3. [](commit_id = e829cd30e3b8f7c1048023ae9980a06a868806f1, deletion_comment = False)

**Comment by pratikjoshi14** [comment]
>     "UserAssignedIdentityProperties": {

Can common type be used here? common-types/resource-management/v4/managedidentity.json#/definitions/UserAssignedIdentity

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:9666 in e829cd3. [](commit_id = e829cd30e3b8f7c1048023ae9980a06a868806f1, deletion_comment = False)

**Comment by hiaga** [comment]
> > ```
> > "RestrictedVaultName": {
> > ```
> 
> Why is this separate path parameter needed? How would it work on existing resources?
> 
> Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:13477 in [e829cd3](https://github.com/Azure/azure-rest-api-specs/commit/e829cd30e3b8f7c1048023ae9980a06a868806f1). [](commit_id = [e829cd3](https://github.com/Azure/azure-rest-api-specs/commit/e829cd30e3b8f7c1048023ae9980a06a868806f1), deletion_comment = False)

@pratikjoshi14 this is added to resolve lint diff error for pattern restriction for new parameters. New type is added so that any of the existing parameters won't be affected by this.

**Comment by pratikjoshi14** [comment]
Every new API might also hit similar issue right? The name ResctrictedVaultName does not look clean in that case.

---
In reply to: [1779044085](https://github.com/Azure/azure-rest-api-specs/pull/25791#issuecomment-1779044085) [](http://example.com/codeflow?ancestors=1779044085)

**Comment by pratikjoshi14** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupTieringCost/default/operationResults/{operationId}": {

We had to add /backupTieringCost/default resource just to return different operation result for /fetchTieringCost. Otherwise these APIs can be moved to tracked resource only.
Check with ARM API modelling team in office hours if this approach is fine or is there any better way to achieve the same.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:4662 in e829cd3. [](commit_id = e829cd30e3b8f7c1048023ae9980a06a868806f1, deletion_comment = False)

**Comment by hiaga** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupTieringCost/default/operationResults/{operationId}": {
> > ```
> 
> We had to add /backupTieringCost/default resource just to return different operation result for /fetchTieringCost. Otherwise these APIs can be moved to tracked resource only. Check with ARM API modelling team in office hours if this approach is fine or is there any better way to achieve the same.
> 
> Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:4662 in [e829cd3](https://github.com/Azure/azure-rest-api-specs/commit/e829cd30e3b8f7c1048023ae9980a06a868806f1). [](commit_id = [e829cd3](https://github.com/Azure/azure-rest-api-specs/commit/e829cd30e3b8f7c1048023ae9980a06a868806f1), deletion_comment = False)

Every new API will rely on this restricted pattern hence avoiding any lint diff failures in future

**Comment by jindani** [comment]
> > > ```
> > > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupTieringCost/default/operationResults/{operationId}": {
> > > ```
> > 
> > 
> > We had to add /backupTieringCost/default resource just to return different operation result for /fetchTieringCost. Otherwise these APIs can be moved to tracked resource only. Check with ARM API modelling team in office hours if this approach is fine or is there any better way to achieve the same.
> > Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2023-06-01/bms.json:4662 in [e829cd3](https://github.com/Azure/azure-rest-api-specs/commit/e829cd30e3b8f7c1048023ae9980a06a868806f1). [](commit_id = [e829cd3](https://github.com/Azure/azure-rest-api-specs/commit/e829cd30e3b8f7c1048023ae9980a06a868806f1), deletion_comment = False)
> 
> Every new API will rely on this restricted pattern hence avoiding any lint diff failures in future

@Heng Lu Suggested us to use this route pattern for our use case.

**Review by rkmanda — APPROVED** [review]
:shipit:

**Pull request merged**